### PR TITLE
metrics: make metrics easier to use with prometheus

### PR DESCRIPTION
**Pull request by cce**
## Summary

Some changes to make it easier to find and use metrics with Prometheus

* report HELP, TYPE, and zero-value counters for all declared Counters, even if they haven't been incremented yet (makes it easier to find them in Grafana/Prometheus, versus waiting for them to be incremented first)
* allow tags used by TagCounter to be pre-declared at initialization time, so even zero-valued TagCounter metrics will show up with HELP and TYPE text and will be seen by Prometheus.
* ensure TagCounter metrics are attributed to a host by handling "parentLabels" when writing Prometheus format
* add counter TYPE line to each TagCounter, allowing our node_exporter to see TagCounters, because our [node_exporter's gateway only handles counter and gauge metrics declared with TYPE](https://github.com/algorand/node_exporter/blob/master/collector/gateway.go#L240-L241).

## Test Plan

New test added to assert Prometheus output from WriteMetric for TagCounter and Counter.

### CODE REVIEW

File: util/metrics/tagcounter.go [hunk 1]
```diff
@@ -98,32 +102,34 @@ func (tc *TagCounter) WriteMetric(buf *strings.Builder, parentLabels string) {
...
 		} else {
-			buf.WriteString(tc.Name)
-			buf.WriteRune('_')
-			buf.WriteString(tag)
-			buf.WriteRune(' ')
-			buf.WriteString(strconv.FormatUint(*tagcount, 10))
-			buf.WriteRune('\n')
+			name = tc.Name + "_" + tag
+		}
+		buf.WriteString("# HELP ")
```
Comment (algonautshant) [inline util/metrics/tagcounter.go:117]: What is the value of adding HELP here?

**Review by algonautshant — APPROVED** [review]
Looks good.

File: util/metrics/tagcounter.go (hunk 1 above)
Comment (cce) [reply util/metrics/tagcounter.go:117]: the HELP line provides a description of the metric for some UIs & backends use ... for example Grafana 6.6+ shows it alongside the metric name when you are browsing metrics https://grafana.com/blog/2020/06/15/how-we-made-working-with-prometheus-easier-with-metric-metadata-in-grafanas-explore-view/

File: util/metrics/counter.go [hunk 2]
```diff
@@ -155,16 +155,23 @@ func (counter *Counter) WriteMetric(buf *strings.Builder, parentLabels string) {
...
 	buf.WriteString(counter.name)
 	buf.WriteString(" counter\n")
+	// if counter is zero, report 0 using parentLabels and no tags
+	if len(counter.values) == 0 {
+		buf.WriteString(counter.name)
+		if len(parentLabels) > 0 {
+			buf.WriteString("{" + parentLabels + "}")
+		}
+		buf.WriteString(" 0")
+		buf.WriteString("\n")
```
Comment (algorandskiy) [inline util/metrics/counter.go:172]: return

**Review by algorandskiy — APPROVED** [review]

**Review by brianolson — APPROVED** [review]

**Comment by brianolson** [comment]
Ok, but now you point out that `.Deregister()` is a good idea, doesn't that mean that _really_ none of the tests should be using any globals anyways?

**Review by brianolson — APPROVED** [review]
Looked a little closer and I see that counter_test.go tests are kinda doing integration, but tagcounter_test.go could be isolated easily
```
-       tc := NewTagCounter("tc", "wat")
+       tc := TagCounter{Name: "tc", Description: "wat"}
```
but, at least the counter_test.go _don't_ call t.Parallel() so they plan on being run serially and isolating the use of the global that way.

**Comment by algorandskiy** [comment]
Why deleted segments. Not used but might be useful on some point...

**Comment by cce** [comment]
I talked with @brianolson and @winder about it and we figured we could bring something conceptually similar back some day in the future if we need it. I think if we were to reimplement it, we could have a sort of timing histogram system that didn't rely on metrics.Counter, perhaps.

**Review by brianolson — APPROVED** [review]

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**